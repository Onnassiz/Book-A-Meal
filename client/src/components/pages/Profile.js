import React, { Component } from 'react';
import ProfileContainer from '../container/ProfileContainer';
import NavBarContainer from '../container/NavBarContainer';

class Profile extends Component {
	render() {
		return (
			<div>
				<NavBarContainer page="caterer" />
				<ProfileContainer />
			</div>
		);
	}
}

export default Profile;
