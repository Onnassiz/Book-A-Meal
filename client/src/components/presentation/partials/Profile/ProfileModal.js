import React from 'react';
import PropTypes from 'prop-types';
import Modal from 'react-modal';
import empty from 'is-empty';
import { addProfileModalStyle } from './../../../../utilities/modalStyles';
import { BasicInput, TextArea } from '../../../presentation/form/BasicInput';
import ShowErrors from './../../partials/ShowErrors';
import SubmitButton from '../../../presentation/form/SubmitButton';

const closeModalStyle = {
  float: 'right',
};

const renderInputFields = data => (
  <div>
    <BasicInput
      name="businessName"
      type="text"
      label="Business Name"
      value={data.state.businessName}
      onChange={data.onChange}
      hasError={data.state.errors.businessName !== undefined}
    />
    <BasicInput
      name="mission"
      type="text"
      label="Mission Statement"
      value={data.state.mission}
      onChange={data.onChange}
      hasError={data.state.errors.mission !== undefined}
    />
    <BasicInput
      name="email"
      type="text"
      label="Business Email"
      value={data.state.email}
      onChange={data.onChange}
      hasError={data.state.errors.email !== undefined}
    />
  </div>
);

const renderModalBody = props => (
  <div>
    <div className="col-12">
      <a onClick={props.handleAddButtonClick} style={closeModalStyle}><i style={{ fontSize: 25 }} className="material-icons">close</i></a>
    </div>
    <div className="box">
      <h3>Setup Business Profile</h3>
      <ShowErrors clientErrors={props.state.errors} serverErrors={props.profile.errors || {}} />
      <form onSubmit={props.handleSubmit}>
        {renderInputFields(props)}
        <TextArea
          name="contact"
          type="text"
          label="Contact"
          value={props.state.contact}
          onChange={props.onChange}
          hasError={props.state.errors.contact !== undefined}
        />
        <SubmitButton value={empty(props.profile.businessName) ? 'Submit' : 'Update'} isLoading={props.formState.isLoading} />
      </form>
    </div>
  </div>
);

const ProfileModal = props => (
  <div>
    <Modal
      isOpen={props.state.isShowingModal}
      closeTimeoutMS={1}
      style={addProfileModalStyle}
      ariaHideApp={false}
      contentLabel="Modal"
      className="create-profile"
    >
      {renderModalBody(props)}
    </Modal>
  </div>
);

ProfileModal.propTypes = {
  state: PropTypes.object.isRequired,
};

renderModalBody.propTypes = {
  state: PropTypes.object.isRequired,
  profile: PropTypes.object.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  formState: PropTypes.func.isRequired,
  handleAddButtonClick: PropTypes.func.isRequired,
};

export default ProfileModal;
