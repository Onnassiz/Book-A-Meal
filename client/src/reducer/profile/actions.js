import axios from 'axios';
import { setLoading, unsetLoading } from '../formState/actions';

export const SET_PROFILE = 'SET_PROFILE';
export const SET_PROFILE_SERVER_ERRORS = 'SET_PROFILE_SERVER_ERRORS';
export const SET_PROFILE_ALERT = 'SET_PROFILE_ALERT';

function setAlert(alert) {
  return dispatch => dispatch({
    type: SET_PROFILE_ALERT,
    alert,
  });
}

export function setProfile(profile) {
  return dispatch => dispatch({
    type: SET_PROFILE,
    id: profile.id,
    businessName: profile.businessName,
    mission: profile.mission,
    contact: profile.contact,
    banner: profile.banner,
    email: profile.email,
  });
}

function setProfileErrors(errors) {
  return dispatch => dispatch({
    type: SET_PROFILE_SERVER_ERRORS,
    errors,
  });
}

export function postProfile(profile) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.post('api/v1/profile', profile).then((response) => {
      dispatch(unsetLoading());
      dispatch(setProfile(response.data));
      dispatch(setAlert('Profile successfully created'));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setProfileErrors(error.response.data));
      return error;
    });
  };
}

export function updateProfile(profile) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.put(`api/v1/profile/${profile.id}`, profile).then((response) => {
      dispatch(unsetLoading());
      dispatch(setProfile(response.data));
      dispatch(setAlert('Profile successfully updated'));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setProfileErrors(error.response.data));
      return error;
    });
  };
}

export function putImage(id, field) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.put(`api/v1/profile/image/${id}`, field).then((response) => {
      dispatch(unsetLoading());
      dispatch(setProfile(response.data));
      dispatch(setAlert('Banner successfully uploaded'));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setProfileErrors(error.response.data));
      return error;
    });
  };
}

export function getProfile() {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.get('api/v1/profile').then((response) => {
      dispatch(unsetLoading());
      dispatch(setProfile(response.data));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      return error;
    });
  };
}
