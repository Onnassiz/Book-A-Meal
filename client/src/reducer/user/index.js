import { SET_USER, SET_SIGN_UP_SERVER_ERRORS, SET_SIGN_IN_SERVER_ERRORS } from './action';

const initialState = {};

export default (state = initialState, action) => {
	switch (action.type) {
	case SET_USER:
		return { name: action.name, email: action.email, role: action.role };
	case SET_SIGN_UP_SERVER_ERRORS:
		return { signUpErrors: action.errors };
	case SET_SIGN_IN_SERVER_ERRORS:
		return { signInErrors: action.errors };
	default:
		return state;
	}
};
