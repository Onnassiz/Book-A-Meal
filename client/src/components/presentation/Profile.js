/* eslint-disable */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import empty from 'is-empty';
import { toast } from 'react-toastify';
import validateProfile from '../../utilities/validateProfileForm';
import ImageUploader from '../presentation/partials/ImageUploader';
import ProfileModal from './partials/Profile/ProfileModal';
import ProfileSection from './partials/Profile/ProfileSection';

class Profile extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isShowingModal: false,
      isShowingDropZoneModal: false,
      mission: '',
      email: props.user.email,
      businessName: '',
      contact: '',
      errors: {},
    };
    this.handleAddButtonClick = this.handleAddButtonClick.bind(this);
    this.toggleDropZone = this.toggleDropZone.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.onChange = this.onChange.bind(this);
    this.putImage = this.putImage.bind(this);
  }

  componentWillMount() {
    const { user, history, getProfile, profile } = this.props;
    if (user.role !== 'caterer' || !user.role) {
      history.push('/');
    }
  }

  componentDidMount() {
    const { getProfile, profile } = this.props;
    document.title = 'Profile - Just Eat';
    if (empty(profile)) {
      getProfile();
    }
  }

  onChange(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  toggleDropZone() {
    this.setState({ isShowingDropZoneModal: !this.state.isShowingDropZoneModal });
  }

  putImage(imageURL) {
    const { putImage, profile } = this.props;
    const field = {
      banner: imageURL,
    };
    putImage(profile.id, field).then((response) => {
      if (response.status === 200) {
        toast('Your profile banner has been updated');
        this.toggleDropZone(false);
      }
    });
  }

  isValid() {
    this.setState({ errors: {} });
    const { errors, isValid } = validateProfile(this.state);
    if (!isValid) {
      this.setState({ errors });
    }
    return isValid;
  }

  handleAddButtonClick(update) {
    if (update) {
      const { profile } = this.props;
      this.setState({
        isShowingModal: !this.state.isShowingModal,
        businessName: profile.businessName,
        email: profile.email,
        mission: profile.mission,
        contact: profile.contact,
      });
    } else {
      this.setState({ isShowingModal: !this.state.isShowingModal });
    }
  }

  createProfile(formData) {
    const { postProfile } = this.props;
    postProfile(formData).then((response) => {
      if (response.status === 201) {
        toast('Your have successfully created a profile');
        this.handleAddButtonClick(false);
      }
    });
  }

  updateProfile(formData) {
    const { updateProfile } = this.props;
    updateProfile(formData).then((response) => {
      if (response.status === 200) {
        toast('Your profile has been updated');
        this.handleAddButtonClick(false);
      }
    });
  }

  handleSubmit(e) {
    e.preventDefault();
    if (this.isValid()) {
      const { profile } = this.props;

      const formData = {
        id: profile.id,
        businessName: this.state.businessName,
        email: this.state.email,
        mission: this.state.mission,
        contact: this.state.contact,
      };

      if (empty(profile.businessName)) {
        this.createProfile(formData);
      } else {
        this.updateProfile(formData);
      }
    }
  }


  render() {
    const { formState, profile } = this.props;
    return (
      <div id="content-body">
        <div className="col-12">
          <button onClick={empty(profile.businessName) ? this.handleAddButtonClick.bind(this, false) : this.handleAddButtonClick.bind(this, true)} style={{ marginRight: 5 }} className="button">
            {empty(profile.businessName) ? 'Add Profile' : 'Update Profile'}
          </button>
          <button disabled={empty(profile.id)} onClick={this.toggleDropZone} className="button">Upload Banner</button>
        </div>

        <ProfileModal
          state={this.state}
          profile={profile}
          handleSubmit={this.handleSubmit}
          handleAddButtonClick={this.handleAddButtonClick}
          formState={formState}
          onChange={this.onChange}
        />
        <div>
          <ImageUploader putImage={this.putImage} isShowingAddPhoto={this.state.isShowingDropZoneModal} toggleAddPhoto={this.toggleDropZone} />
        </div>
        <ProfileSection profile={profile} />
      </div>
    );
  }
}

Profile.propTypes = {
  user: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  profile: PropTypes.object.isRequired,
  formState: PropTypes.object.isRequired,
  postProfile: PropTypes.func.isRequired,
  updateProfile: PropTypes.func.isRequired,
  getProfile: PropTypes.func.isRequired,
  putImage: PropTypes.func.isRequired,
};

export default Profile;
