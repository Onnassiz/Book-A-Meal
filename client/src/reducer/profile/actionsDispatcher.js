import axios from 'axios';
import { setLoading, unsetLoading } from '../formState/actionsDispatcher';
import { setProfileAction, setProfileErrorsAction } from './actions';


export const setProfile = profile => dispatch => dispatch(setProfileAction(profile));

const setProfileErrors = errors => dispatch => dispatch(setProfileErrorsAction(errors));

export const postProfile = profile => (dispatch) => {
  dispatch(setLoading());
  return axios.post('api/v1/profile', profile).then((response) => {
    dispatch(unsetLoading());
    dispatch(setProfile(response.data));
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setProfileErrors(error.response.data));
    return error;
  });
};

export const updateProfile = profile => (dispatch) => {
  dispatch(setLoading());
  return axios.put(`api/v1/profile/${profile.id}`, profile).then((response) => {
    dispatch(unsetLoading());
    dispatch(setProfile(response.data));
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setProfileErrors(error.response.data));
    return error;
  });
};

export const putImage = (id, field) => (dispatch) => {
  dispatch(setLoading());
  return axios.put(`api/v1/profile/${id}`, field).then((response) => {
    dispatch(unsetLoading());
    dispatch(setProfile(response.data));
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setProfileErrors(error.response.data));
    return error;
  });
};

export const getProfile = () => (dispatch) => {
  dispatch(setLoading());
  return axios.get('api/v1/profile').then((response) => {
    dispatch(unsetLoading());
    dispatch(setProfile(response.data));
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setProfileErrors(error.response.data));
    return error;
  });
};
