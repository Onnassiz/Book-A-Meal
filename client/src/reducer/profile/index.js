import { SET_PROFILE, SET_PROFILE_SERVER_ERRORS } from './actions';

const initialState = {};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_PROFILE:
      return {
        id: action.id,
        businessName: action.businessName,
        email: action.email,
        mission: action.mission,
        contact: action.contact,
        banner: action.banner,
      };
    case SET_PROFILE_SERVER_ERRORS:
      return { ...state, errors: action.errors };
    default:
      return state;
  }
};
