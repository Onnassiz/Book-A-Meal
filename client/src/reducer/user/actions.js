import axios from 'axios';
import { setLoading, unsetLoading } from '../formState/actions';

export const SET_USER = 'SET_USER';
export const SET_SIGN_UP_SERVER_ERRORS = 'SET_SIGN_UP_SERVER_ERRORS';
export const SET_SIGN_IN_SERVER_ERRORS = 'SET_SIGN_IN_SERVER_ERRORS';

export function setUser(user) {
  return dispatch => dispatch({
    type: SET_USER,
    id: user.id,
    name: user.name,
    role: user.role,
    email: user.email,
  });
}

function setSignUpErrors(errors) {
  return dispatch => dispatch({
    type: SET_SIGN_UP_SERVER_ERRORS,
    errors,
  });
}

function setSignInErrors(errors) {
  return dispatch => dispatch({
    type: SET_SIGN_IN_SERVER_ERRORS,
    errors,
  });
}

export function postUser(user) {
  return (dispatch) => {
    dispatch(setLoading());
    axios.post('api/v1/auth/signUp', user).then((response) => {
      localStorage.setItem('id', response.data.id);
      localStorage.setItem('email', response.data.email);
      localStorage.setItem('name', response.data.fullName);
      localStorage.setItem('role', response.data.role);
      localStorage.setItem('token', response.data.token);
      setTimeout(() => {
        window.location = '/';
      }, 500);
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setSignUpErrors(error.response.data));
    });
  };
}


export function signInUser(user) {
  return (dispatch) => {
    dispatch(setLoading());
    axios.post('api/v1/auth/signIn', user).then((response) => {
      localStorage.setItem('id', response.data.id);
      localStorage.setItem('email', response.data.email);
      localStorage.setItem('name', response.data.fullName);
      localStorage.setItem('role', response.data.role);
      localStorage.setItem('token', response.data.token);
      setTimeout(() => {
        window.location = '/';
      }, 500);
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setSignInErrors(error.response.data));
    });
  };
}
