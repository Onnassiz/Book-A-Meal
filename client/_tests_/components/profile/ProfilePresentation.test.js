import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Profile from '../../../src/components/presentation/Profile';
import user from '../../objectProps/user.props';
import profile from '../../objectProps/profile.props';
import { postProfile, updateProfile } from '../../promiseProps/profile.props';

user.role = 'customer';
const props = {
  user,
  history: { push: jest.fn() },
  profile: {},
  formState: {},
  postProfile,
  updateProfile,
  getProfile: jest.fn(),
  putImage: updateProfile,
};

const wrapper = shallow(<Profile {...props} />);

describe('<Profile />', () => {
  it('Should render Profile page successfully', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should call onChange method to update handle input change', () => {
    const onChangeSpy = jest.spyOn(wrapper.instance(), 'onChange');
    const firstEvent = {
      target: {
        name: 'businessName',
        value: 'Sweet Summer',
      },
    };
    wrapper.instance().onChange(firstEvent);
    expect(wrapper.state().businessName).toBe('Sweet Summer');
    expect(onChangeSpy).toHaveBeenCalled();
  });

  it('Should trigger handleSubmit method when it is called and call createProfile', () => {
    const handleSubmitSpy = jest.spyOn(wrapper.instance(), 'handleSubmit');
    const createProfileSpy = jest.spyOn(wrapper.instance(), 'createProfile');

    const event = {
      preventDefault: jest.fn(),
    };

    wrapper.setState(profile);

    wrapper.instance().handleSubmit(event);
    expect(handleSubmitSpy).toHaveBeenCalled();
    expect(createProfileSpy).toHaveBeenCalled();
  });

  it('Should call toggleDropZone when button is clicked', () => {
    wrapper.find('#toggleDropZone').simulate('click');
    expect(wrapper.state().isShowingDropZoneModal).toBe(true);
  });

  it('Should trigger handleSubmit method when it is called and call updateProfile', () => {
    const handleSubmitSpy = jest.spyOn(wrapper.instance(), 'handleSubmit');
    const updateProfileSpy = jest.spyOn(wrapper.instance(), 'updateProfile');

    const event = {
      preventDefault: jest.fn(),
    };

    wrapper.setState(profile);

    wrapper.setProps({ profile });

    wrapper.instance().handleSubmit(event);
    expect(handleSubmitSpy).toHaveBeenCalled();
    expect(updateProfileSpy).toHaveBeenCalled();
  });

  it('Should trigger handleSubmit method and throw error when validation fails', () => {
    const handleSubmitSpy = jest.spyOn(wrapper.instance(), 'handleSubmit');

    const event = {
      preventDefault: jest.fn(),
    };

    wrapper.setState({
      businessName: '',
      mission: '',
      contact: '',
      email: '',
      banner: '',
    });

    wrapper.instance().handleSubmit(event);
    expect(wrapper.state().errors.businessName).toBe('The business name field is required');
    expect(wrapper.state().errors.mission).toBe('The mission field is required');
    expect(wrapper.state().errors.email).toBe('The email field is required');
    expect(wrapper.state().errors.contact).toBe('The contact field is required');
    expect(handleSubmitSpy).toHaveBeenCalled();
  });

  it('Should trigger handleAddButtonClick method when it is called', () => {
    const handleAddButtonClickSpy = jest.spyOn(wrapper.instance(), 'handleAddButtonClick');

    wrapper.instance().handleAddButtonClick(true);
    expect(handleAddButtonClickSpy).toHaveBeenCalled();
  });

  it('Should trigger putImage method when it is called', () => {
    const putImageSpy = jest.spyOn(wrapper.instance(), 'putImage');

    profile.businessName = 'Waiters Splash';
    wrapper.setState(profile);

    wrapper.instance().putImage().then(() => {
      expect(putImageSpy).toHaveBeenCalled();
      expect(wrapper.state().isShowingDropZoneModal).toBe(false);
    });
  });
});
