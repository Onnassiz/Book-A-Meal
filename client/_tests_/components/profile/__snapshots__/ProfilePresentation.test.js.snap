// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile /> Should render Profile page successfully 1`] = `
<div
  id="content-body"
>
  <div
    className="col-12"
  >
    <button
      className="button"
      onClick={[Function]}
      style={
        Object {
          "marginRight": 5,
        }
      }
    >
      Add Profile
    </button>
    <button
      className="button"
      disabled={true}
      id="toggleDropZone"
      onClick={[Function]}
    >
      Upload Banner
    </button>
  </div>
  <ProfileModal
    formState={Object {}}
    handleAddButtonClick={[Function]}
    handleSubmit={[Function]}
    onChange={[Function]}
    profile={Object {}}
    state={
      Object {
        "businessName": "",
        "contact": "",
        "email": "onnassiz@gmail",
        "errors": Object {},
        "isShowingDropZoneModal": false,
        "isShowingModal": false,
        "mission": "",
      }
    }
  />
  <div>
    <ImageUploader
      isShowingAddPhoto={false}
      putImage={[Function]}
      toggleAddPhoto={[Function]}
    />
  </div>
  <ProfileSection
    profile={Object {}}
  />
</div>
`;
